### PR TITLE
fix: fix compact physical logic

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
fix validation blocking
* what changes does this pull request make?
fix compact physical logic
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: xline/src/server/kv_server.rs [hunk 1]
```diff
@@ -350,9 +359,33 @@ where
...
+            Either::Right(async {})
+        };
+        let cmd = command_from_request_wrapper(propose_id, wrapper);
+        let (cmd_res, _sync_res) = self
+            .client
+            .propose(cmd, !physical)
+            .await
+            .map_err(client_err_to_status)?;
         let resp = cmd_res.into_inner();
+        compact_physical_fut.await;
```
Comment (iGxnon) [inline xline/src/server/kv_server.rs:388]: Will it be blocked here forever?

Comment (themanforfree) [reply xline/src/server/kv_server.rs]: Added a timeout

File: simulation/src/xline_group.rs [hunk 2]
```diff
@@ -127,3 +169,97 @@ impl Drop for XlineGroup {
         debug!("all nodes shutdowned");
     }
 }
+
+pub struct SimEtcdClient {
```
Comment (bsbds) [inline simulation/src/xline_group.rs:173]: Madsim has a `madsim-etcd-client` crate, could it be used here?

File: xline/src/server/command.rs [hunk 3]
```diff
@@ -70,6 +72,8 @@ where
     general_rev: Arc<RevisionNumberGenerator>,
     /// Revision Number generator for Auth request
     auth_rev: Arc<RevisionNumberGenerator>,
+    /// Compact events
+    compact_events: Arc<RwLock<HashMap<ProposeId, Arc<Event>>>>,
```
Comment (bsbds) [inline xline/src/server/command.rs:76]: We could use `DashMap` here.

File: xline/src/server/command.rs [hunk 4]
```diff
@@ -156,6 +162,16 @@ where
             RequestBackend::Auth => self.auth_storage.after_sync(wrapper, revision)?,
             RequestBackend::Lease => self.lease_storage.after_sync(wrapper, revision).await?,
         };
+        #[allow(clippy::wildcard_enum_match_arm)]
+        let is_physical_compact = match wrapper.request {
+            RequestWrapper::CompactionRequest(ref compact_req) => compact_req.physical,
+            _ => false,
+        };
```
Comment (bsbds) [inline xline/src/server/command.rs:169]: It could use if let.

File: xline/src/server/kv_server.rs [hunk 5]
```diff
@@ -42,30 +47,39 @@ where
     id_barrier: Arc<IdBarrier>,
     /// Range request retry timeout
     range_retry_timeout: Duration,
+    /// Compact timeout
+    compact_timeout: Duration,
     /// Consensus client
     client: Arc<Client<Command>>,
+    /// Compact events
+    compact_events: Arc<RwLock<HashMap<ProposeId, Arc<Event>>>>,
```
Comment (bsbds) [inline xline/src/server/kv_server.rs:55]: Same as Above.

File: xline/src/server/kv_server.rs [hunk 6]
```diff
@@ -42,30 +47,39 @@ where
...
 }
 
 impl<S> KvServer<S>
 where
     S: StorageApi,
 {
     /// New `KvServer`
+    #[allow(clippy::too_many_arguments)]
     pub(crate) fn new(
```
Comment (Phoenix500526) [inline xline/src/server/kv_server.rs:64]: Shall we use the builder pattern to refactor this `new` method?

File: simulation/src/xline_group.rs (hunk 2 above)
Comment (themanforfree) [reply simulation/src/xline_group.rs:173]: `madsim-etcd-client` lacks many methods, such as watch

File: xline/src/server/command.rs (hunk 3 above)
Comment (themanforfree) [reply xline/src/server/command.rs]: Changed

File: xline/src/server/command.rs (hunk 4 above)
Comment (themanforfree) [reply xline/src/server/command.rs]: Fixed

File: xline/src/server/kv_server.rs (hunk 5 above)
Comment (themanforfree) [reply xline/src/server/kv_server.rs]: Changed

File: xline/src/server/kv_server.rs (hunk 6 above)
Comment (themanforfree) [reply xline/src/server/kv_server.rs:64]: This method will only be called once, and the other server implementations are the same, which I think is better if the interface is consistent

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Comment by iGxnon** [comment]
You might need to rebase on master to use `cargo hakari generate` to verify whether the hakari is working normally.

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by bsbds — APPROVED** [review]

**Pull request merged**